moveit_core/utils/src/robot_model_builder.cpp

```diff
@@ -0,0 +1,335 @@
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF AD w/ 2VISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Bryce Willey */
+
+#include <ros/ros.h>
+#include <boost/algorithm/string_regex.hpp>
+#include <boost/math/constants/constants.hpp>
+#include <geometry_msgs/Pose.h>
+#include "moveit/utils/robot_model_builder.h"
+
+namespace moveit
+{
+namespace core
+{
+const std::string LOGNAME = "robot_model_builder";
+moveit::core::RobotModelPtr loadRobot(std::string robot_name)
+{
+  urdf::ModelInterfaceSharedPtr urdf = loadURDF(robot_name);
+  srdf::ModelSharedPtr srdf = loadSRDF(robot_name);
+  moveit::core::RobotModelPtr robot_model;
+  robot_model.reset(new moveit::core::RobotModel(urdf, srdf));
+  return robot_model;
+}
+
+urdf::ModelInterfaceSharedPtr loadURDF(std::string robot_name)
+{
+  boost::filesystem::path res_path(MOVEIT_TEST_RESOURCES_DIR);
+  urdf::ModelInterfaceSharedPtr urdf_model = urdf::parseURDFFile((res_path / robot_name / "urdf/robot.xml").string());
+  if (urdf_model == nullptr)
+  {
+    ROS_ERROR_NAMED(LOGNAME, "Cannot find URDF for %s. Make sure moveit_resources/your robot description is installed",
+                    robot_name.c_str());
+  }
+  return urdf_model;
+}
+
+srdf::ModelSharedPtr loadSRDF(std::string robot_name)
+{
+  boost::filesystem::path res_path(MOVEIT_TEST_RESOURCES_DIR);
+  urdf::ModelInterfaceSharedPtr urdf_model = loadURDF(robot_name);
+  srdf::ModelSharedPtr srdf_model(new srdf::Model());
+  srdf_model->initFile(*urdf_model, (res_path / robot_name / "srdf/robot.xml").string());
+  return srdf_model;
+}
+
+RobotModelBuilder::RobotModelBuilder(std::string name, std::string base_link_name)
+  : urdf_model_(new urdf::ModelInterface()), srdf_writer_(new srdf::SRDFWriter())
+{
+  urdf_model_->clear();
+  urdf_model_->name_ = name;
+
+  urdf::LinkSharedPtr base_link(new urdf::Link);
+  base_link->name = base_link_name;
+  urdf_model_->links_.insert(std::make_pair(base_link_name, base_link));
+
+  srdf_writer_->robot_name_ = name;
+}
+
+void RobotModelBuilder::add(std::string section, std::string type, std::vector<geometry_msgs::Pose> joint_origins)
+{
+  std::vector<std::string> link_names;
+  boost::split_regex(link_names, section, boost::regex("->"));
+  if (link_names.empty())
+  {
+    ROS_ERROR_NAMED(LOGNAME, "No links specified (empty section?)");
+    return;
```

Comment:
if this is going to be used for unit tests it seems the function should return false?